src-self-hosted/link.zig

```diff
@@ -2812,6 +2828,100 @@ pub const File = struct {
+            pub const empty = TextBlock{};
+        };
+
+        pub const SrcFn = struct {
+            pub const empty = SrcFn{};
+        };
+
+        pub fn openPath(allocator: *Allocator, dir: fs.Dir, sub_path: []const u8, options: Options) !*File {
+            assert(options.object_format == .macho);
+
+            const file = try dir.createFile(sub_path, .{ .truncate = false, .read = true, .mode = determineMode(options) });
+            errdefer file.close();
+
+            var macho_file = try allocator.create(MachO);
+            errdefer allocator.destroy(macho_file);
+
+            macho_file.* = openFile(allocator, file, options) catch |err| switch (err) {
+                error.IncrFailed => try createFile(allocator, file, options),
+                else => |e| return e,
+            };
+
+            return &macho_file.base;
+        }
+
+        /// Returns error.IncrFailed if incremental update could not be performed.
+        fn openFile(allocator: *Allocator, file: fs.File, options: Options) !MachO {
+            switch (options.output_mode) {
+                .Exe => {},
+                .Obj => {},
+                .Lib => return error.IncrFailed,
+            }
+            var self: MachO = .{
+                .base = .{
+                    .file = file,
+                    .tag = .macho,
+                    .options = options,
+                    .allocator = allocator,
+                },
+                .ptr_width = switch (options.target.cpu.arch.ptrBitWidth()) {
+                    32 => .p32,
+                    64 => .p64,
+                    else => return error.UnsupportedELFArchitecture,
+                },
+            };
+            errdefer self.deinit();
+
+            // TODO implement reading the macho file
+            return error.IncrFailed;
+            //try self.populateMissingMetadata();
+            //return self;
+        }
+
+        /// Truncates the existing file contents and overwrites the contents.
+        /// Returns an error if `file` is not already open with +read +write +seek abilities.
+        fn createFile(allocator: *Allocator, file: fs.File, options: Options) !MachO {
+            switch (options.output_mode) {
+                .Exe => return error.TODOImplementWritingMachOExeFiles,
+                .Obj => return error.TODOImplementWritingMachOObjFiles,
+                .Lib => return error.TODOImplementWritingLibFiles,
+            }
+        }
+
+        /// Commit pending changes and write headers.
+        pub fn flush(self: *MachO) !void {}
+
+        pub fn deinit(self: *MachO) void {}
+
+        pub fn allocateDeclIndexes(self: *MachO, decl: *Module.Decl) !void {}
+
+        pub fn updateDecl(self: *MachO, module: *Module, decl: *Module.Decl) !void {}
+
+        pub fn updateDeclLineNumber(self: *MachO, module: *Module, decl: *const Module.Decl) !void {}
+
+        /// Must be called only after a successful call to `updateDecl`.
```

Comment:
These docs should probably move to the "base" function so that they don't have to get duplicated on all the implementations